### PR TITLE
Added New functional happy path tests for location controller end-points

### PR DESCRIPTION
**Pull request by ashwini-mv**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PUB-2562

### Change description ###
Added New functional happy path tests for location controller end-points

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/utils/FunctionalTestBase.java [hunk 1]
```diff
@@ -57,6 +58,27 @@ protected Response doPostRequest(final String path, final Map<String, String> ad
...
+            .relaxedHTTPSValidation()
+            .headers(additionalHeaders)
+            .accept("*/*")
+            .multiPart(multiPartKey, multipartFile)
+            .when()
+            .post(path)
+            .thenReturn();
+    }
+
+    protected Response doDeleteRequest(final String path, final Map<String, String> additionalHeaders) {
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/utils/FunctionalTestBase.java:73]: I can see that all the delete endoints in data-management does not have request body. You can keep this delete request but can you remove the one underneath as it is not needed.

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        String filePath = this.getClass().getClassLoader().getResource("location/ValidLocations.csv").getPath();
+        File csvFile = new File(filePath);
+
+        final Response responseUploadLocations = doPostRequestMultiPart(
+            UPLOAD_LOCATIONS_URL,
+            headerMapUploadLocations, LOCATION_LIST, csvFile
+        );
+
+        assertThat(responseUploadLocations.getStatusCode()).isEqualTo(OK.value());
+        assertEquals(responseUploadLocations.as(Location[].class).length, 3,
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:78]: The assertion before this one is using assertJ. This assertion here is using JUnit assert. There are 2 different style of assertions in this test. Can you change this one to user assertJ too like below?

 assertThat(responseUploadLocations.as(Loaction[].class)).hasSize(3);

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        );
+        assertThat(responseGetLocationById.getStatusCode()).isEqualTo(OK.value());
+        assertThat(responseGetLocationById.asString().contains(TEST_LOCATION_NAME));
+
+        final Response responseGetAllLocations = doGetRequest(
+            BASE_LOCATIONS_URL,
+            headerMap
+        );
+        assertThat(responseGetAllLocations.getStatusCode()).isEqualTo(OK.value());
+        assertThat(responseGetAllLocations.asString().contains(TEST_LOCATION_NAME));
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:94]: Using asString() in the assertion does not test the response body properly. that string can be anywhere in the response. It is better to cast the response body as the type returned by the actual endpoint so you can test the location names match up with the location ID:

List<Location> locationResults = Arrays.asList(responseUploadLocations.getBody().as(Location[].class));
assertThat(locationResults.get(0).getLocationId).isEqualTo(2001);
assertThat(locationResults.get(0).getName).isEqualTo(TEST_LOCATION_NAME);
assertThat(locationResults.get(1).getLocationId).isEqualTo(2002);
assertThat(locationResults.get(1).getName).isEqualTo("TestLocationTwo");
assertThat(locationResults.get(2).getLocationId).isEqualTo(2003);
assertThat(locationResults.get(2).getName).isEqualTo("TestLocationThree");

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    private static final String UPLOAD_LOCATIONS_URL = BASE_LOCATIONS_URL + "/upload";
+    private static final String GET_LOCATION_BY_ID_URL = BASE_LOCATIONS_URL + "/" + TEST_LOCATION_ID;
+    private static final String GET_LOCATION_BY_NAME_URL = BASE_LOCATIONS_URL + "/name/" + TEST_LOCATION_NAME
+        + "/language/ENGLISH";
+    private static final String GET_LOCATION_BY_REGION_AND_JURISDICTION = BASE_LOCATIONS_URL + "/filter";
+    private static final String DOWNLOAD_CSV_LOCATIONS_URL = BASE_LOCATIONS_URL + "/download/csv";
+
+    @Test
+    void locationControllerHappyPathTests() throws Exception {
+        Map<String, String> headerMap = new ConcurrentHashMap<>();
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:58]: It is cleaner to use Map.of() in one line when creating Map: 
Map<String, String> headerMap = Map.of(AUTHORIZATION, BEARER + accessToken);

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        assertThat(responseGetAllLocations.asString().contains(TEST_LOCATION_NAME));
+        assertThat(responseGetAllLocations.asString().contains("TestLocationTwo"));
+        assertThat(responseGetAllLocations.asString().contains("TestLocationThree"));
+
+        final Response responseGetLocationByName = doGetRequest(
+            GET_LOCATION_BY_NAME_URL,
+            headerMap
+        );
+        assertThat(responseGetLocationByName.getStatusCode()).isEqualTo(OK.value());
+        assertThat(responseGetLocationByName.asString().contains("2001"));
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:103]: Better to cast to Location object to make sure 2001 is the location ID

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+
+        final Response responseGetLocationByRegionAndJurisdiction =
+            given()
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:107]: You can create a new doGetRequest method on FunctionalTestBase like below: 

protected Response doGetRequest(final String path, final Map<String, String> additionalHeaders, Map<String, String> queryParams) {
        return given()
            .relaxedHTTPSValidation()
            .headers(getRequestHeaders(additionalHeaders))
            .queryParams(queryParams)
            .when()
            .get(path)
            .thenReturn();
    }

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        assertThat(responseDownloadCsv.getStatusCode()).isEqualTo(OK.value());
+        assertThat(responseDownloadCsv.asString().contains(TEST_LOCATION_NAME));
+        assertThat(responseDownloadCsv.asString().contains("TestLocationTwo"));
+        assertThat(responseDownloadCsv.asString().contains("TestLocationThree"));
+
+        final Response responseDeleteLocationById = doDeleteRequest(
+            GET_LOCATION_BY_ID_URL,
+            deleteHeaderMap
+        );
+        assertThat(responseDeleteLocationById.getStatusCode()).isEqualTo(OK.value());
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:132]: When this PR is merged to master, this test is going to run on staging. Don't we need to clean up the locations on staging after the test is run? Currently 2 locations remain

File: src/main/resources/db/migration/V1.11__add_location_reference.sql [hunk 8]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS location_reference (
```
Comment (KianKwa) [inline src/main/resources/db/migration/V1.11__add_location_reference.sql:1]: Currently your PR failed to be deployed in Jenkins. It is likely to be caused by this change. Think we need to discuss this

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/utils/FunctionalTestBase.java (hunk 1 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/utils/FunctionalTestBase.java:73]: Removed the delete request without body

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 2 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:78]: Changed to assertJ.

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 3 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:94]: This returns all courts in staging. Used filter method to assertion.

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 4 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:58]: Used Map.of(

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 6 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:107]: Created new method

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 7 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:132]: Fixed

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java (hunk 5 above)
Comment (ashwini-mv) [reply src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:103]: Fixed.

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        );
+        boolean testCourtTwoReturned = returnedLocations.stream().anyMatch((location) -> TEST_LOCATION_NAME_TWO.equals(
+            location.getName()) && TEST_LOCATION_ID_TWO.equals(location.getLocationId())
+
+        );
+        boolean testCourtThreeReturned = returnedLocations.stream().anyMatch((location) ->
+                                                                                 TEST_LOCATION_NAME_THREE.equals(
+            location.getName()) && TEST_LOCATION_ID_THREE.equals(location.getLocationId())
+        );
+        assertTrue(testCourtOneReturned,"Should return " + TEST_LOCATION_NAME_ONE);
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:126]: You are using the wrong import for AssertJ:
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

Change it to the following import:
import static org.assertj.core.api.Assertions.assertThat;

After that you can simply assert like below by using assertThat + anyMatch directly:

 assertThat(returnedLocations).anyMatch(
            location -> TEST_LOCATION_NAME_ONE.equals(location.getName())
                && TEST_LOCATION_ID_ONE.equals(location.getLocationId())
        );

File: src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        Location returnedLocation = responseGetLocationById.as(Location.class);
+        assertThat(returnedLocation.getLocationId()).isEqualTo(TEST_LOCATION_ID_ONE);
+        assertThat(returnedLocation.getName()).isEqualTo(TEST_LOCATION_NAME_ONE);
+
+        final Response responseGetAllLocations = doGetRequest(
+            BASE_LOCATIONS_URL,
+            headerMap
+        );
+        assertThat(responseGetAllLocations.getStatusCode()).isEqualTo(OK.value());
+        List<? extends Location> returnedLocations = Arrays.asList(responseGetAllLocations.getBody()
```
Comment (KianKwa) [inline src/functionalTest/java/uk/gov/hmcts/reform/pip/data/management/location/LocationTest.java:111]: Does not need 'extends' here, just List<location> is sufficient

**Review by KianKwa — APPROVED** [review]

**Review by ChrisS1512 — APPROVED** [review]

**Pull request merged**